F4 — cascade.pbt.test.ts invariant 3 random seed 회귀

## 발견 시점
chore PR #30 (@fast-check/vitest devDep 제거) 검증 중 `npm test` 실행

## 증상
- spec: `lib/domain/cascade.pbt.test.ts > invariant 3: chainedToPrev=false 첫 발견 후 chain 끊김`
- error: `Property failed after 82 tests` · `Caused by: AssertionError: expected 60000 to be +0`
- @fast-check/vitest 제거와 **무관** — main 의 잠재 결함

## RPN 영향
cascade Critical RPN 80 영역. fix 시점 RPN 재평가 의무.

## Root cause 가설 (5)
1. invariant 3 logic 자체 결함
2. arbXxx generator 가 invariant 와 모순되는 input 생성
3. cascade 함수 자체 bug (60000ms = 1분 timing)
4. generator schedule order chronological 보장 X
5. delta=0 + edited schedule done 조합 같은 edge

## 재시작 절차
1. `cd ~/dev/plan1 && npm test 2>&1 | grep -A5 'Property failed'` → seed 확보
2. `FAST_CHECK_SEED=<seed> npm test lib/domain/cascade.pbt.test.ts` → reproduce
3. shrinker minimal counter-example 분석 (input schedules + editedId + delta)
4. invariant 3 logic 검증 (cascade source 와 비교)
5. 결정: spec fix / source fix / generator 보강

## 관련 문서
- `wiki/projects/plan1/qa-pending.md` § 1 F4 (8 항목 상세)
- `wiki/projects/plan1/risk-matrix.md` § 3 (cascade Critical RPN 80)
- `wiki/shared/llm-coding-traps.md` § 1 분기 동작 환각

## 작성
2026-05-03 KST · Studio 머신 발견

--- a/lib/domain/cascade.pbt.test.ts
+++ b/lib/domain/cascade.pbt.test.ts
@@ -74,6 +74,11 @@ describe('cascade — Property-Based Testing (1000 runs · invariants)', () => {
   })
 
   it('invariant 3: chainedToPrev=false 첫 발견 후 chain 끊김 (그 이후 schedule 들은 shift 안 됨)', () => {
+    // ⚠️ 검증 시점 주의: cascade 는 **shift 전 mutated 의 active sort** 로 chain 판정.
+    //   shift 후 startAt 변경으로 result sort 결과가 다를 수 있음 (특히 동일 startAt 의 schedule 이 stable sort 결과 mutated 와 result 에서 다른 위치).
+    //   → spec 도 mutated (shift 전) sort 기준으로 chain 판정 의무. result sort 기준 검증은 false positive 발생.
+    //   배경: F4 회귀 (PR fix/plan1-cascade-pbt-invariant3-spec-logic · 2026-05-03)
+    //         counter-example: 3 schedule 모두 startAt=0, target chained=false, s-1 chained=true, s-2 chained=false
     fc.assert(
       fc.property(arbSchedules(10), fc.integer({ min: -120, max: 120 }), (schedules, deltaMin) => {
         const target = schedules[0]
@@ -81,25 +86,56 @@ describe('cascade — Property-Based Testing (1000 runs · invariants)', () => {
         const newDur = target.durationMin + deltaMin
         if (newDur < 1) return
 
+        const origEnd = target.startAt + target.durationMin * NS
+        const newEnd = newStart + newDur * NS
+        const delta = newEnd - origEnd
+
         const result = cascade(schedules, target.id, newStart, newDur)
 
-        // active 만 정렬 (cascade 의 동일 로직)
-        const active = result.filter(s => s.status !== 'done').sort((a, b) => a.startAt - b.startAt)
-        const editedIdx = active.findIndex(s => s.id === target.id)
+        // mutated (shift 전) 의 active sort — cascade source 와 동일 로직
+        const mutated = schedules.map(s =>
+          s.id === target.id ? { ...s, startAt: newStart, durationMin: newDur } : s
+        )
+        const mutatedActive = mutated
+          .filter(s => s.status !== 'done')
+          .sort((a, b) => a.startAt - b.startAt)
+        const editedIdx = mutatedActive.findIndex(s => s.id === target.id)
         if (editedIdx === -1) return
 
-        // 첫 unchained schedule 이후의 모든 schedule 은 원본과 startAt 동일해야 함
+        // editedIdx+1 부터 첫 unchained 직전까지: shifted (startAt = original + delta)
+        // 첫 unchained 부터 끝까지: unshifted (startAt = original)
         let chainBroken = false
-        for (let i = editedIdx + 1; i < active.length; i++) {
-          if (!active[i].chainedToPrev) chainBroken = true
+        for (let i = editedIdx + 1; i < mutatedActive.length; i++) {
+          if (!mutatedActive[i].chainedToPrev) chainBroken = true
+          const original = schedules.find(s => s.id === mutatedActive[i].id)!
+          const after = result.find(s => s.id === mutatedActive[i].id)!
           if (chainBroken) {
-            const original = schedules.find(s => s.id === active[i].id)!
-            expect(active[i].startAt).toBe(original.startAt)
+            expect(after.startAt).toBe(original.startAt)
+          } else {
+            expect(after.startAt).toBe(original.startAt + delta)
           }
         }
       }),
-      { numRuns: 500 }
+      { numRuns: 1000 }
     )
+  })
+
+  it('invariant 3 회귀 case (F4 · seed -658135282): 동일 startAt + chained=false target 의 chain 판정', () => {
+    // 회귀 차단 가드 (F4 catch — `wiki/projects/plan1/qa-pending.md` § 1)
+    // 모든 schedule startAt=0 (stable sort 의존) · target s-0 chained=false · s-1 chained=true · s-2 chained=false
+    const schedules: Schedule[] = [
+      { title: '', startAt: 0, durationMin: 1, timerType: 'countup', status: 'pending', chainedToPrev: false, id: 's-0' as ScheduleId, categoryId: 'cat-1', createdAt: 0, updatedAt: 0 },
+      { title: '', startAt: 0, durationMin: 1, timerType: 'countup', status: 'pending', chainedToPrev: true,  id: 's-1' as ScheduleId, categoryId: 'cat-1', createdAt: 0, updatedAt: 0 },
+      { title: '', startAt: 0, durationMin: 1, timerType: 'countup', status: 'pending', chainedToPrev: false, id: 's-2' as ScheduleId, categoryId: 'cat-1', createdAt: 0, updatedAt: 0 },
+    ]
+    const result = cascade(schedules, 's-0' as ScheduleId, 0, 2) // delta = +1*NS
+    const s0 = result.find(s => s.id === 's-0')!
+    const s1 = result.find(s => s.id === 's-1')!
+    const s2 = result.find(s => s.id === 's-2')!
+    expect(s0.startAt).toBe(0)        // target startAt 변경 X
+    expect(s0.durationMin).toBe(2)    // target dur 변경됨
+    expect(s1.startAt).toBe(NS)       // mutated sort i=1 chained=true → shift +60000
+    expect(s2.startAt).toBe(0)        // mutated sort i=2 chained=false → break · unshifted
   })
 
   it('invariant 4: done 상태 schedule 은 shift 영향 받지 않음', () => {

